File: fibonacci_seq/lua/fibonacci_seq.lua
-- Setup Timer Function
local function RunAndTime(func, name, iters)
    local before = os.clock()
    _ = func(iters)
    local after = os.clock()
    print(string.format(name .. " took %0.6f seconds to find fib number " .. iters, after - before))
end

-- recursion
local function FibRec(index)
    if (index == 0) then
        return 0
    elseif (index < 2) then
        return 1
    else
        return FibRec(index - 1) + FibRec(index - 2)
    end
end

print("Fib number 20 is: " .. FibRec(20))
RunAndTime(FibRec, "FibRec", 35)

-- Iterative
local function FibIter(index, Acc, Next)
    if Acc  == nil then Acc = 0 end
    if Next == nil then Next = 1 end

    if (index == 0) then
        return Acc
    else
        return FibIter(index - 1, Next, Acc + Next)
    end
end

print("Fib number 20 is: " .. FibIter(20))
RunAndTime(FibIter, "FibIter", 100000)

File: fibonacci_seq/java/fibonacci_seq.java
import java.util.function.Consumer;
import java.time.Duration;
import java.time.Instant;
class Main {
    public static void main(String[] args) {
        
        // Recursion
        System.out.println("Fib number 20 is: " + FibRec(20));
        RunAndTime(Main::FibRec, "FibRec", 35);

        // Iterative
        System.out.println("Fib number 20 is: " + FibIter(20));
        RunAndTime(Main::FibIter, "FibIter", 100000);

    }

    public static int FibRec(int index) {
        if (index == 0) {
            return 0;
        } else if (index < 2) {
            return 1;
        } else {
            return FibRec(index - 1) + FibRec(index - 2);
        }
    }

    public static int FibIter(int index) {
        return FibIter(index, 0, 1);
    }

    public static int FibIter(int index, int acc, int next) {
        if (index == 0) {
            return acc;
        } else {
            return FibIter(index - 1, next, acc + next);
        }
    }

    public static void RunAndTime(Consumer<Integer> func, String name, int iters) {
        Instant start = Instant.now();
        func.accept(iters);
        Instant end = Instant.now();
        Duration timeElapsed = Duration.between(start, end);
        System.out.println(name + " took " + (timeElapsed.toMillis() / 1000.0) + " seconds to find fib number " + iters);
    }
}